File: backend/crawl/yna.js
const puppeteer = require('puppeteer');

async function crawlYna() {
  const browser = await puppeteer.launch({ headless: true });
  const page = await browser.newPage();

  await page.goto('https://www.yna.co.kr/news?site=navi_latest_depth01', {
    waitUntil: 'networkidle2',
  });

  // 페이지 로드 후 기사 데이터 가져오기
  const result = await page.evaluate(() => {
    const newsList = document.querySelectorAll(
      '#container > div.container521 > div.content03 > div.section01 > section > div > ul > li'
    );
    const currentYear = new Date().getFullYear();
    const extractedData = [];

    newsList.forEach((news) => {
      const titleTag = news.querySelector('div > div > strong > a > span');
      const dateTag = news.querySelector('div > div > span');
      const linkTag = news.querySelector('div > div > strong > a');

      if (titleTag && dateTag && linkTag) {
        // 제목 
        const title = titleTag.innerText.trim();
        // 날짜
        const dateText = dateTag.innerText.trim().split(' ')[0]; // "03-11 15:30" -> "03-11"
        const date = `${currentYear}.${dateText.replace('-', '.')}`; // "2024.03.11"
        // 링크
        const link = linkTag.href.startsWith('http') ? linkTag.href : `https://www.yna.co.kr${linkTag.getAttribute('href')}`;

        extractedData.push({ title, link, date });
      }
    });

    return extractedData;
  });

  await browser.close();

  console.log({ '연합뉴스': result });
  return result;
}

// 약 4초
// crawlYna();
module.exports = crawlYna;

File: backend/crawl/yna.py
import requests
from bs4 import BeautifulSoup
from datetime import datetime


def get_data():
    url = 'https://www.yna.co.kr/news?site=navi_latest_depth01'
    response = requests.get(url)

    # 연도 정보 부재로 현재 연도로 대체(수정 필요)
    current_year = datetime.now().year

    if response.status_code == 200:
        soup = BeautifulSoup(response.text, 'html.parser')

        # ul.list의 li 요소를 선택
        news_list = soup.select('#container > div.container521 > div.content03 > div.section01 > section > div > ul > li')

        result = []
        for news in news_list:
            # 제목 (span 안의 텍스트)
            title_tag = news.select_one('div > div > strong > a > span')
            if title_tag:
                title = title_tag.text.strip()
            else:
                continue

            # 링크 (a 태그의 href)
            link_tag = news.select_one('div > div > strong > a')
            if link_tag and link_tag.attrs.get('href'):
                link = link_tag.attrs['href']
            else:
                continue

            # 날짜 (시간 정보)
            date_tag = news.select_one('div > div > span')
            if date_tag:
                # 시간 (예: 10-30 18:00)
                date_text = date_tag.text.strip()
                # 10-30 -> 2024.10.30
                date_unformatted = date_text.split(' ')[0]
                date = f"{current_year}.{date_unformatted.replace('-', '.')}"
            else:
                continue

            # 결과에 추가
            dict_data = {"title": title, "link": link, "date": date}
            result.append(dict_data)

        return {"연합뉴스": result}
    else:
        print(f"HTTP 요청 실패: {response.status_code}")
        return {"Error": response.status_code}
